.ci/ci-fast-return.sh

```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env bash
+
+# Copyright (c) 2019 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+# Description: Perform CI 'fast path return' checks.
+# Returns: 1 for 'fast return OK', 0 for 'do not fastpath'.
+#
+# Checks for:
+#  - file name patterns from the YAML that will force the CI to run
+#  - file name patterns from the YAML to see if we can skip all files in the PR
+#  - If potentially skipping, looks for a 'force' label on the PR to force the CI
+#    to run anyway
+
+set -e
+
+[ -n "$DEBUG" ] && set -x
+
+cidir=$(dirname "$0")
+source "${cidir}/lib.sh"
+
+# If no branch specified, compare against the master.
+# The 'branch' var is required by the get_pr() lib functions.
+branch=${branch:-master}
+
+# The YAML file containing our filename match patterns.
+yqfile="${cidir}/ci-fast-return.yaml"
+
+# Name of the label, that if set on a PR, will force the CI to be run anyway.
+force_label="force-ci"
+
+# Check if any files changed by the PR either force the CI to run or if
+# we can skip all the files in the PR.
+#
+# Return:
+#  0 - all files are fastpath
+#  1 - at least one non-fastpath file found
+can_we_skip() {
+	# The branch is the baseline - ignore it.
+	[ "$specific_branch" = "true" ] && info "Skip basline branch" >&2 && echo "0" && return
+
+	files=$(get_pr_changed_file_details || true)
+	# Strip off the leading status - just grab last column.
+	files=$(echo "$files"|awk '{print $NF}')
+
+	# no files were changed - I guess we can skip the CI then?
+	[ -z "$files" ] && info "No files found" >&2 && echo "0" && return
+
+	local common_skip=$(yq read ${yqfile} common.skip_patterns)
+	# If nothing is found, yq returns the word 'null' (annoyingley ;-) )
+	[ "$common_skip" == "null" ] && common_skip=""
+	local common_check=$(yq read ${yqfile} common.check_patterns)
+	[ "$common_check" == "null" ] && common_check=""
+
+	# Use just the repo name itself. The YAML does not like having the full
+	# repo path in it.
+	local repo="${kata_repo##*/}"
+	if [ -n "$repo" ]; then
+		local repo_skip=$(yq read ${yqfile} ${repo}.skip_patterns)
+		[ "$repo_skip" == "null" ] && repo_skip=""
+		local repo_check=$(yq read ${yqfile} ${repo}.check_patterns)
+		[ "$repo_check" == "null" ] && repo_check=""
+	else
+		info "No repo set, skipping repo specific patterns" >&2
```

Comment:
Why redirected to `stderr`? Is this a warning? If so, I'd add a `warn()` that writes to `stderr` to minimise the smattering of redirects.

Reply:
The problem there might be that certain CI's will auto-fail if they get any output on `stderr`. What you could do is this instead:

```sh
foo()
{
    local msg="$*"
    info "$msg" >/dev/tty
}

foo "stuff happened"
foo "more exciting things occurred"
```

Reply:
But obviously choose a better name than `foo()` :smile:

Reply:
And I was just about to push an update...
hmm, so, I added the 'warn()' call/func in that you requested so we *did* get something on stderr.... I'll probably revert that back to a `foo()` then.

I reckon with the `{ do_stuff; ret=$?; } || true` idiom, all of this echo stuff to get around `set -e` hurting us is maybe avoidable ... but, that is also a bunch more rewrite work. I'll s/info/foo/g , retry, and push rsn then